packages/data-service/src/user-data.ts

```diff
@@ -0,0 +1,289 @@
+   *
+   * @param {string} filePath - the file path to read.
+   * @returns {Promise<T>} the parsed content.
+   */
+  async read(filePath: string): Promise<T> {
+    const content = (await this._readFile(filePath)).toString();
+    return this._parse(content);
+  }
+
+  async write(filePath: string, object: T): Promise<void> {
+    const content = this._serialize(object);
+    return await this._writeFile(filePath, content);
+  }
+
+  async delete(filePath: string): Promise<void> {
+    const absolutePath = this._resolve(this._basePath, filePath);
+    try {
+      await unlink(absolutePath);
+    } catch (error) {
+      log.error(mongoLogId(1_001_000_001), 'UserData', 'Error deleting file', {
+        path: absolutePath,
+      });
+    }
+  }
+
+  private _defaultParser(this: void, content: string): any {
+    return JSON.parse(content);
+  }
+
+  private _defaultSerializer(this: void, content: unknown): string {
+    return JSON.stringify(content, null, 2);
+  }
+
+  private _conformParseFunctionToThrowOnError<T>(
+    parse: ParseFunction<T>
+  ): ThrowingParseFunction<T> {
+    return (content: string) => {
+      const object = parse(content);
+      if (object === undefined) {
+        const ajvParse = parse as AjvParseFunction<T>;
+
+        throw new Error(
+          `${
+            typeof ajvParse.message === 'string'
+              ? ajvParse.message
+              : 'unknown parse error'
+          } at position: ${ajvParse.position ?? 'unknown'}`
+        );
+      }
+
+      return object;
+    };
+  }
+
+  private _resolve(root: string, filePath: string) {
+    const pathRelativeToRoot = path.relative(root, filePath);
+    const resolved = path.resolve(root, pathRelativeToRoot);
+
+    if (
+      pathRelativeToRoot.startsWith('..') ||
+      path.isAbsolute(pathRelativeToRoot) ||
+      root === resolved
+    ) {
+      throw new Error(
+        `Invalid file path: '${filePath}' is not a subpath of ${root}.`
+      );
+    }
+
+    return resolved;
+  }
+
+  private async _ensureParentDir(filePath: string) {
+    await mkdir(path.dirname(filePath), { recursive: true });
+  }
+
+  private async _readFileAndIgnoreErrors(
+    filePath: string
+  ): Promise<Buffer | undefined> {
+    try {
+      return await readFile(filePath);
+    } catch {
+      log.error(mongoLogId(1_001_000_001), 'UserData', 'Error reading file', {
+        path: filePath,
+      });
+      return undefined;
+    }
+  }
+
+  private async _readFile(filePath: string): Promise<Buffer> {
+    const absolutePath = this._resolve(this._basePath, filePath);
+    return await readFile(absolutePath);
+  }
+
+  private async _readAllWithFileNames(
+    pattern: string
+  ): Promise<Array<{ path: string; content: Buffer }>> {
+    const absolutePattern = this._resolve(this._basePath, pattern);
+
+    const filePaths = await glob(absolutePattern);
+    const filteredFiles = filePaths.filter((path) =>
+      pattern ? new RegExp(pattern).exec(path) : true
```

Comment:
Are we using `pattern` as both a glob expression and a regular expression here? That’s not something that usually works  very well :)

Reply:
Oh no, this is a left over, i was using regexp initially